dns.go

```diff
@@ -0,0 +1,10 @@
+package config
+
+// DNSConfig specifies custom resolvers using DoH
+type DNSConfig struct {
+	// DefaultResolver, if present, is a URL for the default DoH resolver.
+	// If empty, DNS resolution will use the system resolver.
+	DefaultResolver string `json:",omitempty"`
+	// CustomResolvers is a map of domains to URLs for custom DoH resolution.
+	CustomResolvers map[string]string `json:",omitempty"`
```

Comment:
Do we _need_ a default resolver? I'd assume that we could have a single "." entry in a `Resolvers` map, right?

Also, let's briefly (very briefly) document the format (i.e., URL? Multiaddr?).

Reply:
note: some documentation in code can happen here, but the user facing docs live in go-ipfs https://github.com/ipfs/go-ipfs/blob/master/docs/config.md

Reply:
+1  on removing `DefaultResolver` and renaming `CustomResolvers` to  `FQDNResolvers` to be extra clear about what the "custom"  rules are about. (I'd like  to have [FQDN](https://en.wikipedia.org/wiki/Fully_qualified_domain_name) in name so people familiar with DNS will immediately understand format and convention for keys).

## Don't skip in JSON

We should avoid `json:",omitempty"` here. 
It is better for discoverability to have empty `DNS.FQDNResolvers` (ok to have `{}`) – that is what we already do in `Pinning.RemoteServices`. 


### Values 

We could support multiaddrs and URLs, but for DoH  URL will be what people want to copy&paste while setting it up, so ok with saying this needs to be https:// URL with DoH  endpoint.

### Keys

I was thinking about interop with wider DNS ecosystem here and suggest we don't invent any new notation with `*` as I initially syggested, but follow the existing DNS convention of using the [explicit FQDN syntax](https://en.wikipedia.org/wiki/Fully_qualified_domain_name#Syntax) in our config files.

To illustrate:

```json
  "DNS": {
    "FQDNResolvers": {
      "eth.": "https://eth.link-or-cloudflare/path/to/dns-query",
      "crypto.": "https://unstoppablesomething.example.com/path/to/dns-query",
      "libre.": "https://www.opennic.org/path/to/their/dns-query", 
      ".": "https://mozilla.cloudflare-dns.com/dns-query"
    }
  }
```

Idea here is to closely follow the format of [zone files](https://en.wikipedia.org/wiki/Zone_file), which should make things easier to understand.  (The DNS root is unnamed, expressed as the empty label terminated by the dot.)


We should enforce keys are explicit FQDNs by validating them with [`dns.IsFqdn`](https://github.com/miekg/dns/blob/83b388a80c430233c72c45c85517573177b7f716/defaults.go#L269) – removes surface for DNS ambiguity.

Reply:
ok fair enough; we can call the field just `Resolvers` however.

Reply:
I will also make some minor changes in the madns Resolver to normalize to fqdns.